tests/PolyType.Tests/JsonSpecializedTests.cs

```diff
@@ -0,0 +1,155 @@
+    }
+
+    [Theory, PairwiseData]
+    public void RequiredMembersWithPartialCtor_Exhaustive([CombinatorialRange(0, (1 << RequiredMembersWithPartialCtor.FieldCount) - 1)] int mask)
+    {
+        StringBuilder builder = new("{");
+        if ((mask & 0x1) == 0x1) builder.Append("\"A\":\"X\",");
+        if ((mask & 0x2) == 0x2) builder.Append("\"B\":3,");
+        if ((mask & 0x4) == 0x4) builder.Append("\"C\":false,");
+        if ((mask & 0x8) == 0x8) builder.Append("\"D\":false,");
+        if (builder.Length > 1) builder.Length -= 1; // Remove trailing comma
+        builder.Append('}');
+        TestContext.Current.TestOutputHelper?.WriteLine(builder.ToString());
+
+        if ((mask & 0x5) == 0x5)
+        {
+            RequiredMembersWithPartialCtor? actual = AssertRequirementMet<RequiredMembersWithPartialCtor>(builder.ToString());
+            Assert.True(actual is { A: "X", C: false });
+            Assert.Equal((mask & 0x2) == 0x2 ? 3 : 5, actual.B);
+            Assert.Equal((mask & 0x8) == 0x8 ? false : true, actual.D);
+        }
+        else
+        {
+            AssertRequirementsNotMet<RequiredMembersWithPartialCtor>(builder.ToString());
+        }
+    }
+
+    private void AssertRequirementsNotMet<T>([StringSyntax("json")] string json)
+    {
+        var converter = JsonSerializerTS.CreateConverter<T>(providerUnderTest.Provider);
+        JsonException ex = Assert.Throws<JsonException>(() => converter.Deserialize(json));
+        TestContext.Current.TestOutputHelper!.WriteLine(ex.Message);
+    }
+
+    private T? AssertRequirementMet<T>([StringSyntax("json")] string json)
+    {
+        var converter = JsonSerializerTS.CreateConverter<T>(providerUnderTest.Provider);
+        return converter.Deserialize(json);
+    }
+
+    [GenerateShape]
+    public partial record RequiredMembersWithDefaultCtor
+    {
+        internal const int FieldCount = 4;
+
+        public required string A { get; set; }
+        public int B { get; set; } = 5;
+        public required bool C { get; init; } = true;
+        public bool D { get; set; } = true;
+    }
+
+    [GenerateShape]
+    public partial record RequiredMembersWithComprehensiveCtor
+    {
+        internal const int FieldCount = 4;
+
+        [SetsRequiredMembers]
+        public RequiredMembersWithComprehensiveCtor(string a, bool c, int b = 5, bool d = true)
+        {
+            A = a;
+            B = b;
+            C = c;
+            D = d;
+        }
+
+        public required string A { get; set; }
+        public int B { get; set; }
+        public required bool C { get; init; }
+        public bool D { get; set; }
+    }
+
+    [GenerateShape]
+    public partial record RequiredMembersWithPartialCtor
+    {
+        internal const int FieldCount = 4;
+
+        public RequiredMembersWithPartialCtor(string a)
+        {
+            A = a;
+        }
+
+        public string A { get; set; } // TODO: required keyword here causes the JsonSerializer to throw.
```

Comment:
@eiriktsarpalis you might be interested in this. I think this is a pre-existing bug in your sample json serializer. 
Your visitor class throws when I add `required` to this property.